config/servicemappings/compute.yaml

```diff
@@ -469,6 +469,13 @@ spec:
                 version: v1beta1
                 group: compute.cnrm.cloud.google.com
               targetField: self_link
+            - key: serviceAttachmentRef
```

Comment:
Could you add field description?

Reply:
The naming convention is `[originalFieldName]Ref` for a primitive-typed field reference, and `[originalFieldName]Refs` for an array-typed field reference.

https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/README.ConfigureResourceReferences.md#configure-reference-resource-in-the-service-mappings

The reason not to use `[kind]Ref` is that we want to keep naming consistency for single-kind and multi-kind references.

Reply:
`service attachment` is one of the types of `target` filed, and type config does not have "description": https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/53ff50ee2498ddf3e68a5f7e03b4a833cf5957c2/pkg/apis/core/v1alpha1/servicemapping_types.go#L285

Reply:
> When using reference, the field name shall be straightforward and better tell how to use the field.

Agreed, but there are two directions to improve UX:

**Direction I**: How to better tell users which underlying GCP field this field maps to: `[fieldName]Ref`.
Note that the field name has very little to do with the TF design. This is usually just the field name from the GCP API.

Pros:

1. This is important because fundamentally users are using the GCP resources, and knowing which field they are configuring means less learning and better fidelity.
1. This also helps users better understand what the value they should use for the `external` subfield.
1. Most importantly, it provides a consistent interface. If the API decides to support more options for a field, e.g. previously it supports a storage bucket for storage, now it's going to support a spanner database for storage as well, and both are configured under the same field (so user either uses a storage bucket or a spanner database), then we don't need to add a new `[kind]Ref`. Because it's semantically wrong to reference a SpannerDatabase using `storageBucketRef`, using `[kind]ref` resulted in extra cost in such a use case: additional work to our developer, and additional learning/changes to the user. This use case is not the most common changes in the API, but it has resulted multiple suboptimal experiences to several users when we were using `[kind]Ref`. We've learned from the cost and came up with the solution of `[fieldName]Ref`.

Cons:
1. It doesn't tell users directly which kind to reference. Though this information is documented in the CRD.

**Direction II**: How to better tell users which kind to reference: `[kind]Ref`.

Pros:
1. This means users don't need to read the CRD schema to pick the kind.
2. Users don't need to configure `kind` for a reference because there is only one default kind.

Cons:
1. Users still need to know the particular requirement of the kind to be referenced in the field, which means reading the CRD schema/public doc is likely unavoidable - This basically makes the 1st pro not really important.
2. It's difficult for users to understand which underlying GCP field this reference maps to. They may need to do thorough digging before they have a good understanding.
3. As mentioned above, when the API team release a new feature/a new referencable value for the existing reference field, the user needs to learn a new field, the developers need to add a new field to the CRD. This is the additional maintenance cost in the long term.


Besides, @gemmahou could you add field description for this reference?

Reply:
Actually, after re-reading the CRD schema, I see you point, @yuwenma . Yes, this is a multi-kind reference field with the old design that each reference field maps to a specific kind, so when we add a new field referencing a new kind, the field name should be `[kind]Ref`. I think we call it `serviceAttachmentRef` instead of `computeServiceAttachmentRef` just for consistency - the other reference fields also dropped the service `compute` in the field name.

In the future, I'd suggest we support a mutually-exclusive `targetRef` field so that for any newly supported kind under `target`, we should ask users to use `targetRef` instead of adding a `target.[newKind]Ref` field to reference it.

Discussed offline and @gemmahou pointed it out that because it is a multi-kind reference, field description for each reference field is not supported. So we can leave it as is for now. If we've got any feedback about the missing field description, then we can evaluate whether we should improve the UI.